README
NAME
    GraphViz2::Abstract::Edge - Deal with edges independent of a Graph

VERSION
    version 0.001000

SYNOPSIS
        use GraphViz2::Abstract::Edge;

        my $edge = GraphViz2::Abstract::Edge->new(
                color =>  ... ,
                id    =>  ... ,
                label =>  ... ,
        );

        # Mutate $edge

        $edge->label("Asdft");

        my $fillcolor = $edge->fillcolor(); # Knows that the fill color is light grey despite never setting it.

        # Later:

        $graph->add_edge(from => a => to => b => %{ $edge->as_hash }); # Adds only the data that is not the same as GraphViz's defaults
        $graph->add_edge(from => a => to => b => %{ $edge->as_canon_hash }); # Adds all the data, including hardcoded defaults

DESCRIPTION
    Working with GraphViz2, I found myself frequently needing shared styles
    for things, and I often had trouble knowing which fields were and
    weren't valid for given things, for instance: "Edge"s.

    Its reasonably straight forward to ask the question "What is the
    attribute "foo" applicable to" using the GraphViz website, but much
    harder to know "What are all the attributes applicable to "foo"".

    Let alone work with them in a user friendly way from code.

  Naming Rationale
    I tried to choose a name that was not so likely to threaten GraphViz2 if
    GraphViz2 wanted to make a different variation of what I'm doing, but as
    part of GraphViz2 itself.

    As such, I plan on a few "::Abstract" things, that aim to be stepping
    stones for dealing with complex data independent of "GraphViz2", but in
    such a way that they make importing that data into "GraphViz2" easy.

METHODS
  "as_hash"
    This method returns all the values of all properties that DIFFER from
    the defaults.

    e.g.

        Edge->new( color => 'black' )->as_hash();

    Will return an empty list, as the default color is normally black.

    See note about "Special Values"

  "as_canon_hash"
    This method returns all the values of all properties, INCLUDING
    defaults.

    e.g.

        Edge->new( color => 'black' )->as_canon_hash();

    Will return a very large list containing all the properties that we know
    the default values for.

    See note about "Special Values"

ATTRIBUTES
  "URL"
    Default: "none"

  "arrowhead"
    Default: 'normal'

  "arrowsize"
    Default: 0.0

  "arrowtail"
    Default: 'normal'

  "color"
    Default: 'black'

  "colorscheme"
    Default: ""

  "comment"
    Default: ""

  "constraint"
    Default: "true"

  "decorate"
    Default: "false"

  "dir"
    Default: "unknown"

  "edgeURL"
    Default: ""

  "edgehref"
    Default: ""

  "edgetarget"
    Default: "none"

  "edgetooltip"
    Default: ""

  "fillcolor"
    Default: "unknown"

  "fontcolor"
    Default: 'black'

  "fontname"
    Default: 'Times-Roman'

  "fontsize"
    Default: 14.0

  "headURL"
    Default: ""

  "head_lp"
    Default: "unknown"

  "headclip"
    Default: "true"

  "headhref"
    Default: ""

  "headlabel"
    Default: ""

  "headport"
    Default: 'center'

  "headtarget"
    Default: "none"

  "headtooltip"
    Default: ""

  "href"
    Default: ""

  "id"
    Default: ""

  "label"
    Default: ""

  "labelURL"
    Default: ""

  "labelangle"
    Default: -25.0

  "labeldistance"
    Default: 1.0

  "labelfloat"
    Default: "false"

  "labelfontcolor"
    Default: 'black'

  "labelfontname"
    Default: 'Times-Roman'

  "labelfontsize"
    Default: 14.0

  "labelhref"
    Default: ""

  "labeltarget"
    Default: "none"

  "labeltooltip"
    Default: ""

  "layer"
    Default: ""

  "len"
    Default: "unknown"

    Reason: back-end dependent

  "lhead"
    Default: ""

  "lp"
    Default: "unknown"

  "ltail"
    Default: ""

  "minlen"
    Default: 1

  "nojustify"
    Default: "false"

  "penwidth"
    Default: 1.0

  "pos"
    Default: "unknown"

  "samehead"
    Default: ""

  "sametail"
    Default: ""

  "showboxes"
    Default: 0

  "style"
    Default: ""

  "tailURL"
    Default: ""

  "tail_lp"
    Default: "unknown"

  "tailclip"
    Default: "true"

  "tailhref"
    Default: ""

  "taillabel"
    Default: ""

  "tailport"
    Default: 'center'

  "tailtarget"
    Default: "none"

  "tailtooltip"
    Default: ""

  "target"
    Default: "none"

  "tooltip"
    Default: ""

  "weight"
    Default: 1

  "xlabel"
    Default: ""

  "xlp"
    Default: "unknown"

SPECIAL VALUES
    In the specification, on GraphViz.org, there are a number of special
    values which represent a fundamental incompatibility with native Perl
    code.

    *   "false"

        Where the specification shows "false" as a default value, this
        module instead returns the string "false"

        This is because under the hood, GraphViz2 doesn't support values for
        attributes other than defined ones.

        So its assumed that GraphViz, under the hood, interprets the string
        "false" the same as the boolean condition "false";

    *   "true"

        Where the specification shows "true" as a default value, this module
        instead returns the string "true"

        Its assumed that GraphViz, under the hood, interprets the string
        "true" the same as the boolean condition "true", for similar reasons
        "false" is.

    *   "NONE"

        In the GraphViz docs, a few items have a default value specified as:

            <none>

        This item was confusing in the specification, and it wasn't clear if
        it was some sort of magic string, or what.

        Internally, we use a special value, a reference to a string "none"
        to represent this default.

        For instance:

            my $v = Edge->new()->target();

            ok( ref $v, 'target returned a ref' );
            is( ref $v, 'SCALAR', 'target returned a scalar ref' );
            is( ${ $v }, 'none', 'target returned a scalar ref of "none"' );

        However, because its not known how to canonicalize such forms, those
        values are presently not returned by either "as_hash" methods.

        So as a result:

            my $v = Edge->new( color => \"none" )->as_hash()

        Will emit an empty hash. ( Despite "none" being different from the
        default ).

        Also:

            my $v = Edge->new( color => \"none" )->as_canon_hash()

        Will not emit a value for "color" in its output, which may have the
        undesirable effect of reverting to the default, "black" once
        rendered.

    *   "UNKNOWN"

        On the GraphViz documentations, there were quite a few fields where
        the defaults were simply not specified, or their values were
        cryptic.

        Internally, those fields have the default value of "\"unknown""

        Like "none", those fields with those values will not be emitted
        during hash production.

AUTHOR
    Kent Fredric <kentfredric@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Kent Fredric
    <kentfredric@gmail.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for GraphViz2-Abstract-Edge

{{$NEXT}}

0.002000 2013-12-01T06:28:23Z
 - Factor out constants into GraphViz2::Abstract::Util::Constants

 [Dependencies::Stats]
 - Dependencies changed, see Changes.deps{,.all,.dev} for details
 - develop: +2 ↑1 -2
 - runtime: +1 -1

0.001000 2013-11-30T17:03:21Z
 - First version.



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4203"
  },
  "configure_requires" => {
    "Module::Build" => "0.4203"
  },
  "dist_abstract" => "Deal with edges independent of a Graph",
  "dist_author" => [
    "Kent Fredric <kentfredric\@gmail.com>"
  ],
  "dist_name" => "GraphViz2-Abstract-Edge",
  "dist_version" => "0.001000",
  "license" => "perl",
  "module_name" => "GraphViz2::Abstract::Edge",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Class::Tiny" => 0,
    "Scalar::Util" => 0,
    "constant" => 0,
    "perl" => "5.008",
    "strict" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "GraphViz2" => 0,
    "Path::Tiny" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "1.001002",
    "parent" => 0
  }
);


my %fallback_build_requires = (
  "GraphViz2" => 0,
  "Module::Build" => "0.4203",
  "Path::Tiny" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "1.001002",
  "parent" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/GraphViz2/Abstract/Edge.pm
use strict;
use warnings;
use utf8;

package GraphViz2::Abstract::Edge;

# ABSTRACT: Deal with edges independent of a Graph

use GraphViz2::Abstract::Util::Constants;

our @CARP_NOT;

=head1 SYNOPSIS

    use GraphViz2::Abstract::Edge;

    my $edge = GraphViz2::Abstract::Edge->new(
            color =>  ... ,
            id    =>  ... ,
            label =>  ... ,
    );

    # Mutate $edge

    $edge->label("Asdft");

    my $fillcolor = $edge->fillcolor(); # Knows that the fill color is light grey despite never setting it.

    # Later:

    $graph->add_edge(from => a => to => b => %{ $edge->as_hash }); # Adds only the data that is not the same as GraphViz's defaults
    $graph->add_edge(from => a => to => b => %{ $edge->as_canon_hash }); # Adds all the data, including hardcoded defaults

=head1 DESCRIPTION

Working with GraphViz2, I found myself frequently needing shared styles for things, and I often had trouble knowing
which fields were and weren't valid for given things, for instance: C<Edge>s.

Its reasonably straight forward to ask the question "What is the attribute C<foo> applicable to" using the GraphViz website,
but much harder to know "What are all the attributes applicable to C<foo>".

Let alone work with them in a user friendly way from code.

=head2 Naming Rationale

I tried to choose a name that was not so likely to threaten GraphViz2 if GraphViz2 wanted to make a different
variation of what I'm doing, but as part of GraphViz2 itself.

As such, I plan on a few C<::Abstract> things, that aim to be stepping stones for dealing with complex data independent of C<GraphViz2>,
but in such a way that they make importing that data into C<GraphViz2> easy.

=cut

use Class::Tiny {
  URL            => NONE,
  arrowhead      => 'normal',
  arrowsize      => 0.0,
  arrowtail      => 'normal',
  color          => 'black',
  colorscheme    => EMPTY_STRING,
  comment        => EMPTY_STRING,
  constraint     => TRUE,
  decorate       => FALSE,
  dir            => UNKNOWN,
  edgeURL        => EMPTY_STRING,
  edgehref       => EMPTY_STRING,
  edgetarget     => NONE,
  edgetooltip    => EMPTY_STRING,
  fillcolor      => UNKNOWN,
  fontcolor      => 'black',
  fontname       => 'Times-Roman',
  fontsize       => 14.0,
  headURL        => EMPTY_STRING,
  head_lp        => UNKNOWN,
  headclip       => TRUE,
  headhref       => EMPTY_STRING,
  headlabel      => EMPTY_STRING,
  headport       => 'center',
  headtarget     => NONE,
  headtooltip    => EMPTY_STRING,
  href           => EMPTY_STRING,
  id             => EMPTY_STRING,
  label          => EMPTY_STRING,
  labelURL       => EMPTY_STRING,
  labelangle     => -25.0,
  labeldistance  => 1.0,
  labelfloat     => FALSE,
  labelfontcolor => 'black',
  labelfontname  => 'Times-Roman',
  labelfontsize  => 14.0,
  labelhref      => EMPTY_STRING,
  labeltarget    => NONE,
  labeltooltip   => EMPTY_STRING,
  layer          => EMPTY_STRING,
  len            => UNKNOWN,         # backend dependent
  lhead          => EMPTY_STRING,
  lp             => UNKNOWN,
  ltail          => EMPTY_STRING,
  minlen         => 1,
  nojustify      => FALSE,
  penwidth       => 1.0,
  pos            => UNKNOWN,
  samehead       => EMPTY_STRING,
  sametail       => EMPTY_STRING,
  showboxes      => 0,
  style          => EMPTY_STRING,
  tailURL        => EMPTY_STRING,
  tail_lp        => UNKNOWN,
  tailclip       => TRUE,
  tailhref       => EMPTY_STRING,
  taillabel      => EMPTY_STRING,
  tailport       => 'center',
  tailtarget     => NONE,
  tailtooltip    => EMPTY_STRING,
  target         => NONE,
  tooltip        => EMPTY_STRING,
  weight         => 1,
  xlabel         => EMPTY_STRING,
  xlp            => UNKNOWN,
};

=attr C<URL>

Default: L<< C<none>|GraphViz2::Abstract::Util::Constants/NONE >>

=attr C<arrowhead>

Default: C<'normal'>

=attr C<arrowsize>

Default: C<0.0>

=attr C<arrowtail>

Default: C<'normal'>

=attr C<color>

Default: C<'black'>

=attr C<colorscheme>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<comment>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<constraint>

Default: L<< C<true>|GraphViz2::Abstract::Util::Constants/TRUE >>

=attr C<decorate>

Default: L<< C<false>|GraphViz2::Abstract::Util::Constants/FALSE >>

=attr C<dir>

Default: L<< C<unknown>|GraphViz2::Abstract::Util::Constants/UNKNOWN >>

=attr C<edgeURL>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<edgehref>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<edgetarget>

Default: L<< C<none>|GraphViz2::Abstract::Util::Constants/NONE >>

=attr C<edgetooltip>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<fillcolor>

Default: L<< C<unknown>|GraphViz2::Abstract::Util::Constants/UNKNOWN >>

=attr C<fontcolor>

Default: C<'black'>

=attr C<fontname>

Default: C<'Times-Roman'>

=attr C<fontsize>

Default: C<14.0>

=attr C<headURL>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<head_lp>

Default: L<< C<unknown>|GraphViz2::Abstract::Util::Constants/UNKNOWN >>

=attr C<headclip>

Default: L<< C<true>|GraphViz2::Abstract::Util::Constants/TRUE >>

=attr C<headhref>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<headlabel>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<headport>

Default: C<'center'>

=attr C<headtarget>

Default: L<< C<none>|GraphViz2::Abstract::Util::Constants/NONE >>

=attr C<headtooltip>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<href>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<id>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<label>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<labelURL>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<labelangle>

Default: C<-25.0>

=attr C<labeldistance>

Default: C<1.0>

=attr C<labelfloat>

Default: L<< C<false>|GraphViz2::Abstract::Util::Constants/FALSE >>

=attr C<labelfontcolor>

Default: C<'black'>

=attr C<labelfontname>

Default: C<'Times-Roman'>

=attr C<labelfontsize>

Default: C<14.0>

=attr C<labelhref>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<labeltarget>

Default: L<< C<none>|GraphViz2::Abstract::Util::Constants/NONE >>

=attr C<labeltooltip>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<layer>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<len>

Default: L<< C<unknown>|GraphViz2::Abstract::Util::Constants/UNKNOWN >>

Reason: back-end dependent

=attr C<lhead>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<lp>

Default: L<< C<unknown>|GraphViz2::Abstract::Util::Constants/UNKNOWN >>

=attr C<ltail>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<minlen>

Default: C<1>

=attr C<nojustify>

Default: L<< C<false>|GraphViz2::Abstract::Util::Constants/FALSE >>

=attr C<penwidth>

Default: C<1.0>

=attr C<pos>

Default: L<< C<unknown>|GraphViz2::Abstract::Util::Constants/UNKNOWN >>

=attr C<samehead>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<sametail>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<showboxes>

Default: 0

=attr C<style>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<tailURL>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<tail_lp>

Default: L<< C<unknown>|GraphViz2::Abstract::Util::Constants/UNKNOWN >>

=attr C<tailclip>

Default: L<< C<true>|GraphViz2::Abstract::Util::Constants/TRUE >>

=attr C<tailhref>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<taillabel>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<tailport>

Default: C<'center'>

=attr C<tailtarget>

Default: L<< C<none>|GraphViz2::Abstract::Util::Constants/NONE >>

=attr C<tailtooltip>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<target>

Default: L<< C<none>|GraphViz2::Abstract::Util::Constants/NONE >>

=attr C<tooltip>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<weight>

Default: C<1>

=attr C<xlabel>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<xlp>

Default: L<< C<unknown>|GraphViz2::Abstract::Util::Constants/UNKNOWN >>

=cut

use Scalar::Util qw(blessed);
use Scalar::Util qw(refaddr);

sub _is_equal {
  my ( $self, $a_ref, $b_ref ) = @_;

  return   if defined $a_ref     and not defined $b_ref;
  return   if not defined $a_ref and defined $b_ref;
  return 1 if not defined $a_ref and not defined $b_ref;

  ## A and B are both defined.

  return if not ref $a_ref and ref $b_ref;
  return if ref $a_ref and not $b_ref;

  if ( not ref $a_ref and not ref $b_ref ) {
    return $a_ref eq $b_ref;
  }

  ##  A and B are both refs.
  return refaddr $a_ref eq refaddr $b_ref;
}

sub _is_magic {
  my ( $self, $value ) = @_;
  return if not defined $value;
  return if not ref $value;
  my $addr = refaddr $value;
  return 1 if $addr eq refaddr UNKNOWN;
  return 1 if $addr eq refaddr NONE;
  return;
}

sub _foreach_attr {
  my ( $self, $callback ) = @_;
  if ( not blessed($self) ) {
    require Carp;
    local @CARP_NOT = 'GraphViz2::Abstract::Edge';
    Carp::croak('Can\'t call as_hash on a class');
  }
  my $class    = blessed($self);
  my @attrs    = Class::Tiny->get_all_attributes_for($class);
  my $defaults = Class::Tiny->get_all_attribute_defaults_for(__PACKAGE__);
  for my $attr (@attrs) {
    my $value       = $self->$attr();
    my $has_default = exists $defaults->{$attr};
    my $default;
    if ($has_default) {
      $default = $defaults->{$attr};
    }
    $callback->( $attr, $value, $has_default, $default );
  }
  return $self;
}

=method C<as_hash>

This method returns all the values of all properties that B<DIFFER> from the defaults.

e.g.

    Edge->new( color => 'black' )->as_hash();

Will return an empty list, as the default color is normally black.

See also L<< how special constants work in|GraphViz2::Abstract::Util::Constants/CONSTANTS >>

=cut

sub as_hash {
  my ($self) = @_;
  my %output;

  $self->_foreach_attr(
    sub {
      my ( $attr, $value, $has_default, $default ) = @_;
      if ( not $has_default ) {
        return if $self->_is_magic($value);
        $output{$attr} = $value;
        return;
      }
      return if $self->_is_equal( $value, $default );
      return if $self->_is_magic($value);
      $output{$attr} = $value;
    }
  );
  return \%output;
}

=method C<as_canon_hash>

This method returns all the values of all properties, B<INCLUDING> defaults.

e.g.

    Edge->new( color => 'black' )->as_canon_hash();

Will return a very large list containing all the properties that we know the default values for.

See also L<< how special constants work in|GraphViz2::Abstract::Util::Constants/CONSTANTS >>

=cut

sub as_canon_hash {
  my ($self) = @_;
  my %output;
  $self->_foreach_attr(
    sub {
      my ( $attr, $value, $has_default, $default ) = @_;
      return if $self->_is_magic($value);
      $output{$attr} = $value;
    }
  );
  return \%output;

}

1;
